src/compiler/interface.jl

```diff
@@ -171,15 +172,15 @@ const ADictOrGrads = Union{AbstractDict, Grads}
 # Dictionary interface.
 # Don't use the IdDict directly since it may contain some spurious pairs.
 Base.haskey(gs::Grads, x) = x ∈ gs.params 
-Base.keys(gs::Grads) = gs.params
+# Base.keys(gs::Grads) = gs.params
 Base.values(gs::Grads) = (gs.grads[p] for p in gs.params)
```

Comment:
Forwarding `keys` to gs.grads while relying on gs.params for the values is not good. Either we base the dictionary interface entirely on gs.params or we base it on gs.grads, can't have mixed stuff. 

The comments above
```
# Don't use the IdDict directly since it may contain some spurious pairs
```
suggests that we should think changes thoroughly or leave things as they are (which seems the best option to me)

Reply:
What is the spurious stuff you refer to? It contains references to the objects that had gradients along the way.

Reply:
I think Carlo is referring to the existing comment at https://github.com/FluxML/Zygote.jl/pull/823/files#diff-7511b224d7f3ebb56465690de8e307422e3c9798a22bdd4e960d5c86ba6528aaR173. My understanding of that is that `Base.keys(Grads.grads)` may contain items not in `Base.keys(Grads.params)`. This seems like it should never happen though, so is the comment out of date or am I missing some scenario where it could?

Reply:
the comment is not outdated
```julia
julia> using Flux

julia> m = Chain(Dense(2,2), x->relu.(x), BatchNorm(2)) 
Chain(Dense(2, 2), #1, BatchNorm(2))

julia> gs = gradient(() -> sum(m(rand(2,2))), Flux.params(m))
Grads(...)

julia> gs.grads
IdDict{Any, Any} with 8 entries:
  Float32[0.0, 0.0]                              => [0.0, 0.0]
  BatchNorm(2)                                   => RefValue{Any}((λ = nothing, β = nothing, γ = nothing, μ = nothing, σ² = nothing, ϵ = 0.0, momentum = nothing, affi…
  Float32[-0.63824 0.222623; -0.785237 0.536415] => [0.0 0.0; 0.0 0.0]
  :(Main.m)                                      => (layers = (nothing, nothing, RefValue{Any}((λ = nothing, β = nothing, γ = nothing, μ = nothing, σ² = nothing, ϵ = …
  Box([0.0; 0.0])                                => RefValue{Any}((contents = nothing,))
  Float32[0.0, 0.0]                              => [2.0, 2.0]
  Box([0.0; 0.0])                                => RefValue{Any}((contents = nothing,))
  Float32[1.0, 1.0]                              => [0.0, 0.0]
```

Reply:
we have to keep the current dict interface based on `gs.params`

Reply:
Then per Carlo's point, `Base.values(gs::Grads)` should also forward to `.grads` as well. Having the 2 be differently sized is unexpected (i.e. potentially subtly breaking), and arguably breaking the contract of `keys` + `values`.

Reply:
Yes, I'll add that